lib/DxilPIXPasses/DxilPIXMeshShaderOutputInstrumentation.cpp

```diff
@@ -0,0 +1,374 @@
+
+  auto *PreviousValue = BC.Builder.CreateCall(
+      AtomicOpFunc,
+      {
+          AtomicBinOpcode, // i32, ; opcode
+          m_OutputUAV,      // %dx.types.Handle, ; resource handle
+          AtomicAdd, // i32, ; binary operation code : EXCHANGE, IADD, AND, OR,
+                     // XOR, IMIN, IMAX, UMIN, UMAX
+          Zero32Arg, // i32, ; coordinate c0: index in bytes
+          UndefArg,  // i32, ; coordinate c1 (unused)
+          UndefArg,  // i32, ; coordinate c2 (unused)
+          Increment, // i32); increment value
+      },
+      "UAVIncResult");
+
+  return BC.Builder.CreateAnd(PreviousValue, m_OffsetMask, "MaskedForUAVLimit");
+}
+
+Value *DxilPIXMeshShaderOutputInstrumentation::writeDwordAndReturnNewOffset(
+    BuilderContext &BC, Value *TheOffset, Value *TheValue) {
+
+  Function *StoreValue =
+      BC.HlslOP->GetOpFunc(OP::OpCode::BufferStore, Type::getInt32Ty(BC.Ctx));
+  Constant *StoreValueOpcode =
+      BC.HlslOP->GetU32Const((unsigned)DXIL::OpCode::BufferStore);
+  UndefValue *Undef32Arg = UndefValue::get(Type::getInt32Ty(BC.Ctx));
+  Constant *WriteMask_X = BC.HlslOP->GetI8Const(1);
+
+  (void)BC.Builder.CreateCall(
+      StoreValue,
+      {StoreValueOpcode, // i32 opcode
+       m_OutputUAV,       // %dx.types.Handle, ; resource handle
+       TheOffset,        // i32 c0: index in bytes into UAV
+       Undef32Arg,       // i32 c1: unused
+       TheValue,
+       Undef32Arg, // unused values
+       Undef32Arg, // unused values
+       Undef32Arg, // unused values
+       WriteMask_X});
+
+  m_RemainingReservedSpaceInBytes -= sizeof(uint32_t);
+  assert(m_RemainingReservedSpaceInBytes >=
+         0); // or else the caller didn't reserve enough space
+
+  return BC.Builder.CreateAdd(
+      TheOffset,
+      BC.HlslOP->GetU32Const(static_cast<unsigned int>(sizeof(uint32_t))));
+}
+
+bool DxilPIXMeshShaderOutputInstrumentation::runOnModule(Module &M) {
+  DxilModule &DM = M.GetOrCreateDxilModule();
+  LLVMContext &Ctx = M.getContext();
+  OP *HlslOP = DM.GetOP();
+
+  Instruction *firstInsertionPt =
+      dxilutil::FirstNonAllocaInsertionPt(DM.GetEntryFunction());
+  IRBuilder<> Builder(firstInsertionPt);
+
+  BuilderContext BC{M, DM, Ctx, HlslOP, Builder};
+
+  m_OffsetMask = BC.HlslOP->GetU32Const(UAVDumpingGroundOffset() - 1);
+
+  m_OutputUAV = addUAV(BC);
+
+  auto FlattenedGroupId = insertInstructionsToCalculateFlattenedGroupId(BC);
+  auto FlattenedThreadId = insertInstructionsToCalculateFlattenedThreadId(BC);
+
+  auto F = HlslOP->GetOpFunc(DXIL::OpCode::EmitIndices, Type::getVoidTy(Ctx));
+  auto FunctionUses = F->uses();
+  for (auto FI = FunctionUses.begin(); FI != FunctionUses.end();) {
+    auto &FunctionUse = *FI++;
+    auto FunctionUser = FunctionUse.getUser();
+
+    auto Call = cast<CallInst>(FunctionUser);
+
+    IRBuilder<> Builder2(Call);
+    BuilderContext BC2{M, DM, Ctx, HlslOP, Builder2};
+
+    {
+      constexpr uint32_t DwordCount = 1   // index indicator
```

Comment:
As a suggestion, you could have a helper method, e.g.,

template <typename... T>
void Instrument(BuilderContext &BC, T... values)
{
    llvm::SmallVector<llvm::Value *, 10> Values({static_cast<llvm::Value *>(T)...});
    const uint32_t DwordCount = Values.size();
    llvm::Value * byteOffset = reserveDebugEntrySpace(BC, DwordCount * sizeof(uint323_t);
    for (llvm::Value *V : Values)
    {
        byteOffset = writeDwordAndReturnNewOffset(BC, byteOffset, V);
    }
}

Reply:
That's quite elegant. I'll do so. Thanks.